core/templates/dev/head/domain/topic/TopicValidationService.js

```diff
@@ -0,0 +1,179 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to validate the consistency of a topic. These
+ * checks are performable in the frontend to avoid sending a potentially invalid
+ * topic to the backend, which performs similar validation checks to these
+ * in topic_domain.Topic and subsequent domain objects.
+ */
+
+oppia.factory('TopicValidationService', [
+  'TopicObjectFactory', 'SubtopicObjectFactory', 'SkillSummaryObjectFactory',
+  function(
+      TopicObjectFactory, SubtopicObjectFactory, SkillSummaryObjectFactory) {
+    var _validateSubtopic = function(subtopic) {
+      var issues = [];
+      if (!(subtopic instanceof SubtopicObjectFactory)) {
+        issues.push('All subtopics should be Subtopic objects');
+        return issues;
+      }
+      if (typeof subtopic.getId() !== 'number') {
+        issues.push('Subtopic id should be a number');
+      }
+      if (typeof subtopic.getTitle() !== 'string') {
+        issues.push('Subtopic title should be a string');
+      }
+      if (subtopic.getSkillSummaries().constructor !== Array) {
```

Comment:
You know, thinking about it -- I think we can remove this check (and other "is array" checks). It's not going to come up in practice, right?

Basically, what we should validate here is stuff that the user can do to get things into a bad state -- we should catch that before it goes to the backend. But if there's no chance of it happening at all (I can't see how the list of skill summaries would never be a list, for example) then we can skip that check. Then if a backend error happens, that tells us there's some serious error in our application anyway, and that's working as intended.

But validation errors caused by users should be guarded against and never reach the backend. Also, another sort of check that seems useful is stuff where you can imagine it running into problems due to developer error in the future. For that sort of thing, I think it's fine to have a guard here too. I think subtopic.getId() and "no duplicate stories in canonical story IDs" fall into this category.

Anyway, just some thoughts. Not really against having the "!== Array" check here but it seems slightly overkill and could distract from the more relevant checks.

Reply:
EDIT: 

Looking further down the PR, it looks like the point of these checks are to show validation errors to users.

In that case, I think we should handle these two types of things differently:
- For stuff that the user can fix manually, do what you've done here.
- For stuff that it's impossible for the user to fix manually, raise an Error straightaway. Maybe ask the user to report the bug to admin@oppia.org with a screenshot of the page? Or at the very least do a $log.error() (so it gets reported in the backend) and then throw.
  - In fact, for stuff that's along the lines of "is X the correct type of object", maybe that shouldn't be caught here at all. It's not something the user can fix. We should just throw/raise an error if it's detected, and ultimately the backend will do that anyway. So I think we don't need those checks as part of this PR.

Reply:
Ok, so basically, anything that can be tested in the unit tests, I'll have it as an issue. Rest all (the is variable typeof checks), I will have it throw an Error.

Reply:
SG. Thanks!

Reply:
Done, so actually going through the editor, granted everything works, the only validation error that the user would see is if topic name is empty, or if canonical and additional stories list have some stories common. Made those and some more validations as issues, and made all the constructor checks to throw Error.